services/src/main/java/org/keycloak/protocol/oidc/utils/OIDCRedirectUriBuilder.java

```diff
@@ -204,12 +223,20 @@ public OIDCRedirectUriBuilder addParam(String paramName, String paramValue) {
 
         @Override
         public Response build() {
+            RealmModel realm = client.getRealm();
+
+            responseJWT.issuer(Urls.realmIssuer(session.getContext().getUri().getBaseUri(), realm.getName()));
+            responseJWT.audience(client.getClientId());
+            responseJWT.exp((long) (Time.currentTime() + realm.getAccessCodeLifespan()));
+
             if(clientSession != null) {
-                responseJWT.issuer(clientSession.getNote(OIDCLoginProtocol.ISSUER));
```

Comment:
I think that in case that clientSession is not null, the issuer should be still kept to `clientSession.getNote(OIDCLoginProtocol.ISSUER);` . The issuer based on `Urls.realmIssuer(session.getContext().getUri().getBaseUri(), realm.getName())` should be used just in case that clientSession is null.

I believe this is the reason of the test failures, which look like regression of this PR (EG. DefaultHostnameTest).

Reply:
Changed.